### PR TITLE
Validate Bind Address

### PR DESCRIPTION
**Pull request by wyc**
In reference to issues:
https://github.com/apache/couchdb/issues/896

- We want to ensure that bind_address is actually valid before using it
  to configure a node.

P.S., I "learned" Erlang about an hour ago, so please be extra harsh. Thanks.

### CODE REVIEW

**Comment by janl** [comment]
> P.S., I "learned" Erlang about an hour ago, so please be extra harsh. Thanks.

@wyc this looks great! I’ll do a more thorough review tomorrow.

> In reference to issues:
> apache/couchdb#896
> apache/couchdb#897

Generally, we’d make this two separate commits, just because they are different intents If you fancy, feel free to push a version where these are separate commits. But no biggie yet.

**Comment by janl** [comment]
and ignore travis for now :)

File: src/setup.erl [hunk 1]
```diff
@@ -30,15 +30,29 @@ require_node_count(undefined) ->
...
-error_bind_address() ->
-    throw({error, "Cluster setup requires bind_addres != 127.0.0.1"}).
-
-require_bind_address("127.0.0.1", undefined) ->
-    error_bind_address();
-require_bind_address("127.0.0.1", <<"127.0.0.1">>) ->
-    error_bind_address();
-require_bind_address(_, _) ->
-    ok.
+error_local_bind_address() ->
```
Comment (janl) [inline src/setup.erl:33]: 👍 nice clearer function name

File: src/setup.erl [hunk 2]
```diff
@@ -30,15 +30,29 @@ require_node_count(undefined) ->
...
+    throw({error, "Cluster setup requires a remote bind_address (not 127.0.0.1 nor ::1)"}).
+
+require_remote_bind_address(OldBindAddress, NewBindAddress) ->
```
Comment (janl) [inline src/setup.erl]: 👍 nice clearer function name

File: src/setup.erl [hunk 3]
```diff
@@ -30,15 +30,29 @@ require_node_count(undefined) ->
...
+    case {OldBindAddress, NewBindAddress} of
+        {"127.0.0.1", undefined} -> error_local_bind_address();
+        {_, <<"127.0.0.1">>} -> error_local_bind_address();
+        {"::1", undefined} -> error_local_bind_address();
+        {_, <<"::1">>} -> error_local_bind_address();
+        {_, undefined} -> ok;
+        {_, PresentNewBindAddress} -> require_valid_bind_address(PresentNewBindAddress)
+    end.
```
Comment (janl) [inline src/setup.erl:48]: 🤔 While this definitely works and I’m not requesting a change here, I’d like to take the opportunity to explain the pattern that is used on the right.

Erlang allows for function signature overloading. I.e. the same function can have different function bodies depending on the arguments you pass in. That’s what’s used in the old version of the code. In your version, you use a `case` construct that does a functionally equivalent thing. Each matching condition(the thing before the `->` could be a function argument, and the clause (the thing after the `->` could be the function body.

That is a little more idiomatic Erlang. It doesn’t make much of a difference for this code, but imagine a longer function with multiple nested case statements, you end up with a lot of indentation real quickly, while the function matching pattern keeps the indentation to a minimum and the code more readable.

File: src/setup.erl [hunk 4]
```diff
@@ -30,15 +30,29 @@ require_node_count(undefined) ->
...
+
+error_invalid_bind_address(InvalidBindAddress) ->
```
Comment (janl) [inline src/setup.erl]: 🎨 would maybe move up to the other `error_*` function

**Review by janl — APPROVED** [review]
👍 nice work, especially for a first timer in this code and new to Erlang. Thanks!

**Comment by wyc** [comment]
@janl thanks for the thorough review of my first contribution.

As you suggested, I moved the log message changes out to here:
https://github.com/apache/couchdb-setup/pull/16

**Comment by janl** [comment]
Fixing in https://github.com/apache/couchdb/pull/1439 @wyc can you close this PR? Thanks!

**Comment by wyc** [comment]
Sure thing, thank you @janl!

**Pull request closed without merging**